File: leetcode/dynamic_programming/linear/10. Regular Expression Matching/solution.cpp
class Solution {
public:
    bool isMatch(string s, string p) {
        int m = s.length();
        int n = p.length();

        const int N = 25;
        bool dp[N][N];
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                dp[i][j] = false;
            }
        }

        dp[0][0] = true;

        for (int i = 1; i <= n; i++) {
            if (p[i - 1] == '*')
                dp[0][i] = dp[0][i - 2];
        }

        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (p[j - 1] != '*')
                    dp[i][j] = (s[i - 1] == p[j - 1] || p[j - 1] == '.') && dp[i - 1][j - 1];
                else {
                    dp[i][j] = dp[i][j - 2];
                    dp[i][j] = dp[i][j] || ((s[i - 1] == p[j - 2] || p[j - 2] == '.') && dp[i - 1][j]);
                }
                // cout << i << " " << j << " " << dp[i][j] << endl;
            }
        }
        return dp[m][n];
    }
};

File: leetcode/dynamic_programming/linear/10. Regular Expression Matching/solution.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        m, n = len(s), len(p)
        dp = [[False] * (n + 1) for _ in range(m + 1)]
        dp[0][0] = True
        for i in range(1, n + 1):
            if p[i - 1] == '*' and i > 1:
                dp[0][i] = dp[0][i - 2]

        for i in range(1, m + 1):
            for j in range(1, n + 1):
                if p[j - 1] != '*':
                    dp[i][j] = (s[i - 1] == p[j - 1] or p[j - 1] == '.') and dp[i - 1][j - 1]
                if p[j - 1] == '*' and j > 1:
                    dp[i][j] = dp[i][j - 2] # * means 0
                    dp[i][j] = dp[i][j] or ((s[i - 1] == p[j - 2] or p[j - 2] == '.') and dp[i - 1][j])
            
        return dp[m][n]